enhancements/machine-api/control-plane-machinesets.md

```diff
@@ -0,0 +1,227 @@
+  - https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/auto-cert-recovery.md
+  - https://github.com/openshift/machine-config-operator/blob/master/docs/etcd-quorum-guard.md
+  - https://github.com/openshift/cluster-kube-scheduler-operator
+  - https://github.com/openshift/cluster-kube-controller-manager-operator
+  - https://github.com/openshift/cluster-openshift-controller-manager-operator
+  - https://github.com/openshift/cluster-kube-apiserver-operator
+  - https://github.com/openshift/cluster-openshift-apiserver-operator
+replaces:
+superseded-by:
+---
+
+# Managing Control Plane machines
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+For clarity in this doc we set the definition for "Control Plane" as "The collection of stateless and stateful processes which enable a Kubernetes cluster to meet minimum operational requirements". This includes: kube-apiserver, kube-controller-manager, kube-scheduler, kubelet and etcd.
+
+This proposal outlines a solution for presenting the Control Plane compute as a single scalable resource (i.e machineSet) for each failure domain. Particularly:
+- Proposes to let the installer create a MachineSet for each master Machine.
+- Proposes to let the installer create a Machine Health Check resource to monitor the Control Plane Machines.
+
+**Note:**
+- This will enable to create a new CRD to manage the Control Plane as a single scalable resource on top of MachineSets, autoscaling and fully automated rolling upgrades if desired in the near future. This is out of the scope for this particular proposal.
+
+- A separate proposal is created to let the installer create Machine Health Checking resources to monitor the worker Machines. This is out of the scope for this particular proposal.
+
+## Motivation
+
+The Control Plane is the most critical and sensitive entity of a running cluster. Today OCP Control Plane instances are "pets" and therefore fragile. There are multiple scenarios where adjusting the compute capacity which is backing the Control Plane components might be desirable either for resizing or repairing.
+
+Currently there is nothing that automates or eases this task. The steps for the Control Plane to be resized in any manner or to recover from a tolerable failure are completely manual. Different teams are following different "Standard Operating Procedure" documents scattered around with manual steps resulting in loss of information, confusion and extra efforts for engineers and users.
+
+### Goals
+
+- To have a declarative mechanism to scale Control Plane compute resources.
```

Comment:
use "vertically" here because horizontal scaling isn't really a key goal.

Reply:
> use "vertically" here because horizontal scaling isn't really a key goal.

@smarterclayton I think horizontal scaling is the explicit goal here.

Reply:
Not sure how that got mixed up, but the specific ask here (as the instigator of this) is to allow us to resize masters.  We do *not* have a goal to scale to more than 3 masters.  If horizontally scaling (growing # masters to more than 3) in order to maintain quorum while a master is rotated out to change it's instance type is part of the design, then I think that's fine, but horizontal scaling in and of itself is *not* an explicit ask from the SRE teams.

Reply:
Updated to clarify that horizontal is out of the scope, It's an impl detail that we might choose to leverage in the future.